File: 0064_minimum_path_sum.py
# https://leetcode-cn.com/problems/minimum-path-sum/

class Solution:
    def minPathSum(self, grid):
        """
        Inputs:
            grid: list[list[int]]
        Outputs:
            res: int
        """
        
        m = len(grid)
        n = len(grid[0])
        dp = [0 for _ in range(n)]
        
        # Initialization.
        dp[0] = grid[0][0]
        for j in range(1,n):
            dp[j] = dp[j-1] + grid[0][j]
            
        # Dynamic programming.
        for i in range(1,m):
            dp[0] = dp[0] + grid[i][0]
            for j in range(1,n):
                dp[j] = min(dp[j-1], dp[j]) + grid[i][j]
        
        res = dp[-1]
        return res

File: Cpp/0064_minimum_path_sum.cpp
// https://leetcode.cn/problems/minimum-path-sum/

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int rows = grid.size(), cols = grid[0].size();
        vector<int> dp(cols, grid[rows-1][cols-1]);

        for (int col = cols-2; col >= 0; --col) {
            dp[col] = grid[rows-1][col] + dp[col+1];
        }

        for (int row = rows-2; row >= 0; --row) {
            dp[cols-1] = grid[row][cols-1] + dp[cols-1];
            for (int col = cols-2; col >= 0; --col) {
                dp[col] = grid[row][col] + min(dp[col], dp[col+1]);
            }
        }

        return dp[0];
    }
};
